packages/malloy/src/tags.ts

```diff
@@ -0,0 +1,177 @@
+ * including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import {Annotation} from './model';
+
+type MalloyTagProperties = Record<string, string | boolean>;
+interface PropertyTag {
+  properties: MalloyTagProperties;
+}
+interface DocTag {
+  docString: string;
+}
+type MalloyTag = DocTag | PropertyTag;
+function isDocTag(a: MalloyTag | undefined): a is DocTag {
+  return (a as DocTag)?.docString !== undefined;
+}
+
+export interface MalloyTags extends PropertyTag {
+  docStrings: string[];
+}
+
+export interface Taggable {
+  getTags: () => Tags;
+}
+
+/**
+ * Collection of useful functions for handing tag data ...
+ */
+export class Tags {
+  private tags: Annotation | undefined;
+  constructor(annotation: Annotation | undefined) {
+    this.tags = annotation;
+  }
+
+  getMalloyTags(): MalloyTags {
+    const ret: MalloyTags = {docStrings: [], properties: {}};
+    for (const tagLine of tagList(this.tags)) {
+      const tag = parseTag(tagLine, ret.properties);
+      if (isDocTag(tag)) {
+        ret.docStrings.push(tag.docString);
+      }
+    }
+    return ret;
+  }
+
+  getTagList(): string[] {
+    return tagList(this.tags);
+  }
+}
+
+function tagList(tagNode: Annotation | undefined): string[] {
+  if (!tagNode) {
+    return [];
+  }
+  return [
+    ...(tagNode.inherits ? tagList(tagNode.inherits) : []),
+    ...(tagNode.blockNotes || []),
+    ...(tagNode.notes || []),
+  ];
+}
+
+/**
+ * Lines which start '#"' are doc strings
+ * Lines which start '# ' are property lines ... The property parser is
+ * pretty simple, the langauge is:
+ * Property can be
+ *   name       -- set the value to "true"
+ *   -name      -- delete a value from the property
+ *   name=value -- assign value to the name
+ * Name can be any sequence of non space characters, or a string
+ * A value is any sequence of non space characters, or a string
+ * A string is " enclosing and ending in " with \" allowed inside
+ */
+function parseTag(
+  src: string,
+  tagProp: MalloyTagProperties
+): MalloyTag | undefined {
+  if (src.startsWith('#" ')) {
+    return {docString: src.slice(3)};
+  }
+  if (!src.startsWith('# ')) {
```

Comment:
Looks like our default parser won't allow `#dashboard` (vs `# dashboard`). I think this might be confusing to people? I guess we can make the syntax highlighter highlight the former in a way that makes it clear it's wrong...

Reply:
Yeah, because I am, in my mind, imagining the character after the '#' determines who gets the annotation in a world where there are multiple users, and i didn't want to own all '#[a-z]' annotations

Reply:
Yep, makes sense. I'm good with that